tests/expected/matlab/DefaultFuncInt.m

```diff
@@ -1,6 +1,12 @@
 function varargout = DefaultFuncInt(varargin)
       if length(varargin) == 2 && isa(varargin{1},'numeric') && isa(varargin{2},'numeric')
         functions_wrapper(8, varargin{:});
+      elseif length(varargin) == 1 && isa(varargin{1},'numeric')
+        functions_wrapper(9, varargin{:});
+      elseif length(varargin) == 1 && isa(varargin{1},'numeric')
```

Comment:
Isn't it a C++ syntax thing where optional arguments always need to go after required arguments? Doesn't matlab do the same thing?

Reply:
matlab is weird since it's left to the function implementation to decide how to parse arguments, so you can pretty much do whatever you want w.r.t. optional arguments.  Though it is true that there will be many of these if-blocks that are impossible to reach, but I figured the added flexibility with optional arguments was worth the useless if-blocks.

I think I left a comment in the recursion step of `expand_default_arguments` saying "disable this expand to disallow skipping arguments" or something like that which would change the behavior to what you propose.

Reply:
But then how would you disambiguate those two if conditions? If I call `DefaultFuncInt(b)`, it would still hold true for the first elseif block and never go to the next elseif block. Or am I missing something here?

Reply:
I think just always taking precedence for the first argument being specified and the second skipped (vs first skipped and specifying second) is a reasonable policy which is implemented currently.

i.e. the `DefaultFuncInt(default, b)` will never get called (aka impossible to reach).

A truly matlab-ic way to do things would be allow empty lists to be placeholders (instead of boost::none), e.g.
`DefaultFuncInt([], b)`.  I think this is probably best put into a different PR, but I guess I could implement that in this PR if you want? (in additional to some other normal overloading rules).

I mean basically the only difference between what you're proposing and what I've implemented is that, for example,

`func(int a = 0, string s = "")`

can be called as `func("non-default string")` in this implementation whereas it cannot be called that way in your more strict parsing proposal.  Everything else is the same.

The only caveat is if there happens to be a different, custom function overload `func(string s)` in which case overload resolution will depend on the order the functions were defined in the .i file, but then that seems a little bit dubious anyway.

Reply:
> I think just always taking precedence for the first argument being specified and the second skipped (vs first skipped and specifying second) is a reasonable policy which is implemented currently.
> 
> i.e. the `DefaultFuncInt(default, b)` will never get called (aka impossible to reach).

I am not sure how this is a reasonable policy. If I'm a matlab user, you're telling me for `DefaultFuncInt(int a, int b)` that `DefaultFuncInt(a)` and `DefaultFuncInt(b)` are both the same underlying call, when my intention is clearly different. Moreover this is would add bloat to the wrapper and make it confusing for the next person who comes along to update this. If a particular code path is never accessed, it shouldn't exist.

> I mean basically the only difference between what you're proposing and what I've implemented is that, for example,
> 
> `func(int a = 0, string s = "")`
> 
> can be called as `func("non-default string")` in this implementation whereas it cannot be called that way in your more strict parsing proposal. Everything else is the same.
> 
> The only caveat is if there happens to be a different, custom function overload `func(string s)` in which case overload resolution will depend on the order the functions were defined in the .i file, but then that seems a little bit dubious anyway.

This caveat is important because users of the wrapper wouldn't know about the underlying semantics of which overload is being called. If we specify overloaded functions in the .i file in a different order, we shouldn't see different behavior, right? That's just going to make using the wrapper painful.

Reply:
By forcing users to call `func(0, "non-default string")` from their Matlab script also makes their code more readable, makes our job of maintaining the wrapper easier, and removes all doubt about ambiguities, which is why I am suggesting implementing the C++-like strategy rather than the matlab-like strategy. It's a really cool feature, but it seems prone to lots of errors and the matlab wrapper is a pain to maintain anyway.

Reply:
ok let me change it to the c++ way

Reply:
Just fyi @varunagrawal I just realized there's a test that's illegal in c++ so I changed that.